src/controller/tests/TestAutoCommissioner.cpp

```diff
@@ -685,7 +781,281 @@ TEST_F(AutoCommissionerTest, IsSecondaryNetworkSupportedCombinations)
                          c.name, result, c.isSecondaryNetworkSupported, c.supportsConcurrent, c.hasWiFiCreds, c.hasThreadDataset,
                          c.wifiEndpoint, c.threadEndpoint);
         }
-        EXPECT_EQ(result, c.isSecondaryNetworkSupported);
+        EXPECT_EQ(result, c.isSecondaryNetworkSupported) << c.name;
     }
 }
+
+TEST_F(AutoCommissionerTest, GetNextCommissioningStageNetworkSetup_ReturnsCorrectStageAndError)
+{
+    struct Case
+    {
+        const char * name;
+        bool supportsConcurrent;
+        bool trySecondary;
+        bool hasWiFiCreds;
+        bool hasThreadDataset;
+        EndpointId wifiEndpoint;
+        EndpointId threadEndpoint;
+        CommissioningStage expectedStage;
+        CHIP_ERROR expectedError;
+    };
+
+    Case cases[] = {
+        { .name               = "SecondaryNetworkAtRoot_ExpectThread",
+          .supportsConcurrent = true,
+          .trySecondary       = true,
+          .hasWiFiCreds       = true,
+          .hasThreadDataset   = true,
+          .wifiEndpoint       = kRootEndpointId,
+          .threadEndpoint     = kRootEndpointId,
+          .expectedStage      = CommissioningStage::kThreadNetworkSetup,
+          .expectedError      = CHIP_NO_ERROR },
+
+        { .name               = "PrimaryNetworkWithSecondary_ExpectWiFi",
+          .supportsConcurrent = true,
+          .trySecondary       = false,
+          .hasWiFiCreds       = true,
+          .hasThreadDataset   = true,
+          .wifiEndpoint       = kRootEndpointId,
+          .threadEndpoint     = kRootEndpointId,
+          .expectedStage      = CommissioningStage::kWiFiNetworkSetup,
+          .expectedError      = CHIP_NO_ERROR },
+
+        { .name               = "WiFiOnly_ExpectWiFi",
+          .supportsConcurrent = false,
+          .trySecondary       = false,
+          .hasWiFiCreds       = true,
+          .hasThreadDataset   = false,
+          .wifiEndpoint       = kRootEndpointId,
+          .threadEndpoint     = kInvalidEndpointId,
+          .expectedStage      = CommissioningStage::kWiFiNetworkSetup,
+          .expectedError      = CHIP_NO_ERROR },
+
+        { .name               = "ThreadOnly_ExpectThread",
+          .supportsConcurrent = false,
+          .trySecondary       = false,
+          .hasWiFiCreds       = false,
+          .hasThreadDataset   = true,
+          .wifiEndpoint       = kInvalidEndpointId,
+          .threadEndpoint     = kRootEndpointId,
+          .expectedStage      = CommissioningStage::kThreadNetworkSetup,
+          .expectedError      = CHIP_NO_ERROR },
+
+        { .name               = "NoNetworks_ExpectCleanupError",
```

Comment:
no wifi and no thread with no endpoint IDs for either implies an ethernet device, right? So you're relying on this function not being called in the state machine for such devices?